server/replication/replication_mode.go

```diff
@@ -396,78 +411,158 @@ func (m *ModeManager) tickDR() {
+		}
+	case drStateAsyncWait:
+		if canSync {
 			m.drSwitchToSync()
+			break
+		}
+		if oldAvailableStores := m.drGetAvailableStores(); !reflect.DeepEqual(oldAvailableStores, stores[primaryUp]) {
+			m.drSwitchToAsyncWait(stores[primaryUp])
+			break
+		}
+		if m.drCheckStoreStateUpdated(stores[primaryUp]) {
+			m.drSwitchToAsync(stores[primaryUp])
+		}
+	case drStateAsync:
+		if canSync {
+			m.drSwitchToSyncRecover()
+			break
+		}
+		if !reflect.DeepEqual(m.drGetAvailableStores(), stores[primaryUp]) && m.drCheckStoreStateUpdated(stores[primaryUp]) {
+			m.drSwitchToAsync(stores[primaryUp])
+		}
+	case drStateSyncRecover:
+		if !canSync {
+			m.drSwitchToAsync(stores[primaryUp])
 		} else {
-			m.updateRecoverProgress(progress)
+			m.updateProgress()
+			progress := m.estimateProgress()
+			drRecoverProgressGauge.Set(float64(progress))
+
+			if progress == 1.0 {
+				m.drSwitchToSync()
+			} else {
+				m.updateRecoverProgress(progress)
+			}
 		}
 	}
 
 	m.checkReplicateFile()
 }
 
-func (m *ModeManager) checkStoreStatus() (primaryDownCount, drDownCount, primaryUpCount, drUpCount int) {
+const (
+	primaryUp = iota
+	primaryDown
+	drUp
+	drDown
+	storeStatusTypeCount
+)
+
+func (m *ModeManager) checkStoreStatus() [][]uint64 {
 	m.RLock()
 	defer m.RUnlock()
+	stores := make([][]uint64, storeStatusTypeCount)
 	for _, s := range m.cluster.GetStores() {
-		down := !s.IsTombstone() && s.DownTime() >= m.config.DRAutoSync.WaitStoreTimeout.Duration
+		if s.IsTombstone() {
+			continue
+		}
+		down := s.DownTime() >= m.config.DRAutoSync.WaitStoreTimeout.Duration
 		labelValue := s.GetLabelValue(m.config.DRAutoSync.LabelKey)
 		if labelValue == m.config.DRAutoSync.Primary {
 			if down {
-				primaryDownCount++
+				stores[primaryDown] = append(stores[primaryDown], s.GetID())
 			} else {
-				primaryUpCount++
+				stores[primaryUp] = append(stores[primaryUp], s.GetID())
 			}
 		}
 		if labelValue == m.config.DRAutoSync.DR {
 			if down {
-				drDownCount++
+				stores[drDown] = append(stores[drDown], s.GetID())
 			} else {
-				drUpCount++
+				stores[drUp] = append(stores[drUp], s.GetID())
 			}
 		}
 	}
-	return
+	for i := range stores {
+		sort.Slice(stores[i], func(a, b int) bool { return stores[i][a] < stores[i][b] })
+	}
+	return stores
+}
+
+// UpdateStoreDRStatus saves the dr-autosync status of a store.
+func (m *ModeManager) UpdateStoreDRStatus(id uint64, status *pb.StoreDRAutoSyncStatus) {
+	m.drStoreStatus.Store(id, status)
+}
+
+func (m *ModeManager) drGetAvailableStores() []uint64 {
+	m.RLock()
+	defer m.RUnlock()
+	return m.drAutoSync.AvailableStores
+}
+
+func (m *ModeManager) drCheckStoreStateUpdated(stores []uint64) bool {
+	state := m.GetReplicationStatus().GetDrAutoSync()
+	for _, s := range stores {
+		status, ok := m.drStoreStatus.Load(s)
+		if !ok {
+			return false
+		}
+		drStatus := status.(*pb.StoreDRAutoSyncStatus)
+		if drStatus.GetState() != state.GetState() || drStatus.GetStateId() != state.GetStateId() {
```

Comment:
Which state of the store should be in `AsyncWait`?

Reply:
`AsyncWait` is the intermediate state from `sync` to `async`. (the [chart](https://github.com/tikv/pd/blob/656ce79a6c4bbacc6b7300453359ead6be7e2595/server/replication/replication_mode.go#L440-L456) shows how states change) We can recover dr replica from `AsyncWait` state, it will not have ACID issue.

Reply:
once the TiKV switch to `AsyncWait`, what tikv's behavior looks like？

Reply:
Yes. Though it is not implemented in tikv yet. Now TiKV only switches states.